### PR TITLE
Doc on structuring scripts

### LINKED ISSUE #334: Docs for structuring scripts
We also would like to include documentation on structuring scripts. 

For now, we can include a either a section on the notebook docs page (#135/#136), or a standalone page that links to the notebooks page, with this info:

- Follow the overall structure/order presented for notebooks, except without the markdown or yaml.
- How to include session info information.

### PR DESCRIPTION
**Pull request by sjspielman**
### Which issue does this address?

Closes #334

### Briefly describe the scope of the added docs file, including whether you link out to any external docs.

I've taken a first stab here at docs for structuring scripts, and got some stub links along the way (including one more I realized we could fill in now). I tried to keep things minimal without repeating the notebook docs too much, but let me know if more repetition would be better! Also I was a bit hesitant to include too many relative links to the notebook docs since I wonder if some of the subheaders will change once jupyter gets added.

Something else I thought of while writing this was whether we actually want to split things into "writing code" and "structuring your files", for example if we want to point specifically to docs about conventions like `::` without having to point to notebook docs. 

Either way, I'm not sure if I got the overall flow and organization right here, so please let me know what you think the next steps are for this doc!

### Will you need additional visual aids for these docs?

No

### Author checklist

- [x] The docs page has been added to the `nav` section in `mkdocs.yml`
- [x] The docs adhere to the style guide [**structure directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#structure)
- [x] The docs adhere to the style guide [**language directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#structure)
- [x] The docs adhere to the style guide [**tone directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#tone)
- [x] The docs follow additional contributing instructions in [the docs contributing file](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/contributing.md)

### Reviewer checklist

Please refer to the [docs style guide](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md) and the [docs contributing guide](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/contributing.md) while reviewing this pull request.

- [ ] Do the docs adhere to the style guide [**structure directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#structure)?
- [ ] Do the docs adhere to the style guide [**language directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#structure)?
- [ ] Do the docs adhere to the style guide [**tone directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#tone)?
- [ ] Do the docs follow additional contributing instructions in [the docs contributing file](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/contributing.md)?

### CODE REVIEW

**Review by jashapiro — COMMENTED** [review]
This is a good start, but I think it should probably be a bit more script-specific. 

In particular, I might modify some of the section headings, and I would include some suggestions to use an argument parser (`argparse` for Python or `optparse` for R) to handle input and output paths. 

I debating using `argparse` in the template Python script, but I was hesitant to make it too complicated. I think it might be worth doing though. 

(Similarly, I also debated whether to use a `main()` function as is usual python convention, but decided against it at the time. Part of me wants to revisit that, but I don't think it is necessarily worth it.)

**Comment by sjspielman** [comment]
> (Similarly, I also debated whether to use a main() function as is usual python convention, but decided against it at the time. Part of me wants to revisit that, but I don't think it is necessarily worth it.)

I meant to say - I had thought about this but ended up not including anything because I didn't love the external links I found about it! Let's see how things fall out during my revisions.

File: .github/components/dictionary.txt [hunk 1]
```diff
@@ -35,6 +35,7 @@ GFM
 GitHub
 GitKraken
 GitKraken's
+hardcode
```
Comment (jashapiro) [inline .github/components/dictionary.txt:38]: Let's use ["hard code"](https://en.wikipedia.org/wiki/Hard_coding) or "hard-code"
```suggestion
```

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+### Considerations for any kind of script
+
+- The first line of your R script should be a ["shebang,"](https://linuxhandbook.com/shebang/) which stands for "ha**sh** (`#`) **bang** (`!`)".
+    - This makes the script executable by giving the exact path to the interpreter to use when running the script.
+    - Below, we provide specific instructions for how to write this line in [R scripts](#r-shebang) and [Python scripts](#python-shebang).
+- Your opening comments should be a brief, 1-3 sentences describing the purpose of the script.
+    - These comments _are complementary to_ but do not replace [additional documentation in your module's `README.md`](documenting-analysis.md).
+- If you need to import a separate script with reusable functions, you should import it in the section where you load other packages.
+- In most cases, your script should not hardcode input and output file paths.
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:34]:
```suggestion
- In most cases, your script should not hard code input and output file paths.
```

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+Scripts are always a great place to add code or functions you tend to use frequently.
+
+This page covers our recommendations for how you can structure your analysis scripts, which can store in the [`scripts` folder of your analysis module](./index.md#skeleton-analysis-module-contents).
+
+To learn more about writing a wrapper script to run all scripts and/or notebooks in a module, please refer to [this documentation on running analysis modules](running-a-module.md).
+
+
+Similar to our [recommendations for computational notebooks](notebook-structure.md), we recommend using a similar overall structure in your scripts to enhance sharing and reproducibility:
+
+<!-- Do we want to recommend a shebang? -->
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:13]:
```suggestion
```

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+- **[Shebang](https://linuxhandbook.com/shebang/):** Add an opening line to set the script's interpreter.
+- **Opening comments**: Add comments to the top of your script that briefly explain the script.
+- **Load packages**: Load required packages.
+- **Functions**: Define any custom functions used in the script.
+- **Parse input arguments**: Parse input arguments, such as file paths, here.
+- **Main body**: This is where the main body of your script should appear.
+- [Optional] **Session info**: Export out the versions of all packages used in the analysis.
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:20]: Maybe?
```suggestion
- **Session info** (optional): Export out the versions of all packages used in the analysis.
```

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+
+Similar to our [recommendations for computational notebooks](notebook-structure.md), we recommend using a similar overall structure in your scripts to enhance sharing and reproducibility:
+
+<!-- Do we want to recommend a shebang? -->
+- **[Shebang](https://linuxhandbook.com/shebang/):** Add an opening line to set the script's interpreter.
+- **Opening comments**: Add comments to the top of your script that briefly explain the script.
+- **Load packages**: Load required packages.
+- **Functions**: Define any custom functions used in the script.
+- **Parse input arguments**: Parse input arguments, such as file paths, here.
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:18]:
```suggestion
- **Parse script arguments**: Parse any arguments to the script, such as file paths, here.
```

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+Your R scripts should always start with this line:
+
+```r
+#!/usr/bin/env Rscript
+```
+
+#### Parse input arguments { #r-arguments }
+
+We recommend using the [`optparse` package](https://cran.r-project.org/web/packages/optparse/index.html) to parse input arguments in R scripts.
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:52]: CRAN says to use these kinds of links
```suggestion
We recommend using the [`optparse` package](https://cran.r-project.org/package=optparse) to parse input arguments in R scripts.
```

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+For example, after loading the `optparse` package, you can define input options with the [`make_option()`](https://rdrr.io/cran/optparse/man/add_make_option.html) function, and then use the [`parse_args()`](https://rdrr.io/cran/optparse/man/parse_args.html) function to parse them into a usable list.
+
+```r
+# Set up arguments
+option_list <- list(
+  make_option(
+    # Users can specify this argument with either `-i` or `--input_rds_file`
+    opt_str = c("-i", "--input_rds_file"),
+    # The provided value will be stored as a character variable called `input_anndata_file`
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:62]:
```suggestion
    # The provided value will be stored as a character variable called `input_rds_file`
```

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    help = "path to input RDS file for analysis"
+  ),
+  ##  define additional options here ##
+)
+
+# Parse arguments into a usable list
+opt_list <- parse_args(OptionParser(option_list = option_list))
+```
+
+In the above example, you can access the parsed argument as `opt_list$input_rds_file`.
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:75]: I think being less verbose here is worth the cost, as `option_list` and `opt_list` are way too similar.
````suggestion
opts <- parse_args(OptionParser(option_list = option_list))
```

In the above example, you can access the parsed argument as `opts$input_rds_file`.
````

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    "--input_anndata_file",
+    # The provided value will be stored as a string called `input_anndata_file`
+    type = "character",
+    dest = "input_adddata_file",
+    # Description of the argument
+    help = "path to input AnnData file for analysis"
+)
+```
+
+In the above example, you can access the parsed argument as `parser.input_anndata_file`.
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:117]: Missing a step here:
````suggestion
# define any additional arguments here

# parse the arguments into an object
args = parser.parse_args()
```

In the above example, you can access the parsed argument as `args.input_anndata_file`.
```

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+For example, after loading the `argparse` package, you can define a parser object with [`argparse.ArgumentParser()`](https://docs.python.org/3/library/argparse.html#creating-a-parser), and then use [`.add_argument()`](https://docs.python.org/3/library/argparse.html#adding-arguments) to add each argument to the parser object.
+
+
+```python
+# define a parser object with helpful information for script users
+parser = argparse.ArgumentParser(
+    description = "Performs task XYZ as part of analysis"
+)
+
+# parse a command line argument
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:104]:
```suggestion
# define a command line argument
```

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+```python
+#!/usr/bin/env python3
+```
+
+#### Parse input arguments { #python-arguments }
+
+We recommend using the [`argparse` package](https://docs.python.org/3/library/argparse.html) to parse input arguments in Python scripts.
+
+For example, after loading the `argparse` package, you can define a parser object with [`argparse.ArgumentParser()`](https://docs.python.org/3/library/argparse.html#creating-a-parser), and then use [`.add_argument()`](https://docs.python.org/3/library/argparse.html#adding-arguments) to add each argument to the parser object.
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:95]:
```suggestion
For example, after loading the `argparse` package, you can define a parser object with [`argparse.ArgumentParser()`](https://docs.python.org/3/library/argparse.html#creating-a-parser), and then use the [`.add_argument()` method](https://docs.python.org/3/library/argparse.html#adding-arguments) to add each argument to the parser object.
Finally, use the [`.parse_args()` method](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.parse_args) to create an object containing the argument values.
```

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+To learn more about writing a wrapper script to run all scripts and/or notebooks in a module, please refer to [this documentation on running analysis modules](running-a-module.md).
+
+
+Similar to our [recommendations for computational notebooks](notebook-structure.md), we recommend using a similar overall structure in your scripts to enhance sharing and reproducibility:
+
+<!-- Do we want to recommend a shebang? -->
+- **[Shebang](https://linuxhandbook.com/shebang/):** Add an opening line to set the script's interpreter.
+- **Opening comments**: Add comments to the top of your script that briefly explain the script.
+- **Load packages**: Load required packages.
+- **Functions**: Define any custom functions used in the script.
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:17]: I don't know if we want this here or in the "considerations" section.
```suggestion
- **Functions**: Define any custom functions used in the script.
All functions should include documentation comments.
```

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    help = "path to input AnnData file for analysis"
+)
+```
+
+In the above example, you can access the parsed argument as `parser.input_anndata_file`.
+
+
+#### Additional Python tips
+
+- You may wish to enclose your opening comments inside a [triple-quote block](https://www.geeksforgeeks.org/triple-quotes-in-python/).
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:122]: I don't know that this is worth including? (Also, I am inordinately annoyed by the curly quotes in their syntax block) But I might include something about docstrings specifically: https://peps.python.org/pep-0257/ or https://www.geeksforgeeks.org/python-docstrings/

**Review by jashapiro — COMMENTED** [review]
Getting close. A few fixes and some suggestions for updates. I think we should emphasize a bit more about documentation here, especially for functions.

File: docs/contributing-to-analyses/analysis-modules/script-structure.md [hunk 14]
```diff
@@ -112,13 +117,19 @@ parser.add_argument(
...
-In the above example, you can access the parsed argument as `parser.input_anndata_file`.
+In the above example, you can access the parsed argument as `args.input_anndata_file`.
+
 
 
 #### Additional Python tips
 
-- You may wish to enclose your opening comments inside a [triple-quote block](https://www.geeksforgeeks.org/triple-quotes-in-python/).
-- Following Python convention, you may also want to define [a `main()` function](https://realpython.com/python-main-function/) to handle most of the script work.
+- Please use [docstrings](https://www.geeksforgeeks.org/python-docstrings/) when adding documentation to any custom functions you write.
```
Comment (jashapiro) [inline docs/contributing-to-analyses/analysis-modules/script-structure.md:132]: Make this a bit more direct? Or maybe something in between. I mostly don't want it to sound like documentation is optional.
```suggestion
- Include [docstrings](https://www.geeksforgeeks.org/python-docstrings/) for all custom functions you write.
```

**Review by jashapiro — APPROVED** [review]
LGTM, with only one small comment.

**Pull request merged**